core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -0,0 +1,198 @@
+use std::{future::Future, time::Duration};
+
+use anyhow::Context;
+use chrono::{NaiveDateTime, Utc};
+use rand::Rng;
+use tokio::sync::watch;
+use zksync_config::DADispatcherConfig;
+use zksync_da_client::{types::DAError, DataAvailabilityClient};
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_types::L1BatchNumber;
+
+use crate::metrics::METRICS;
+
+#[derive(Debug)]
+pub struct DataAvailabilityDispatcher {
+    client: Box<dyn DataAvailabilityClient>,
+    pool: ConnectionPool<Core>,
+    config: DADispatcherConfig,
+}
+
+impl DataAvailabilityDispatcher {
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: DADispatcherConfig,
+        client: Box<dyn DataAvailabilityClient>,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            client,
+        }
+    }
+
+    pub async fn run(self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        loop {
+            if *stop_receiver.borrow() {
+                break;
+            }
+
+            tokio::join!(
+                async {
+                    if let Err(err) = self.dispatch().await {
+                        tracing::error!("dispatch error {err:?}");
+                    }
+                },
+                async {
+                    if let Err(err) = self.poll_for_inclusion().await {
+                        tracing::error!("poll_for_inclusion error {err:?}");
+                    }
+                }
+            );
+
+            if tokio::time::timeout(self.config.polling_interval(), stop_receiver.changed())
+                .await
+                .is_ok()
+            {
+                break;
+            }
+        }
+
+        tracing::info!("Stop signal received, da_dispatcher is shutting down");
+        Ok(())
+    }
+
+    /// Dispatches the blobs to the data availability layer, and saves the blob_id in the database.
+    async fn dispatch(&self) -> anyhow::Result<()> {
+        let mut conn = self.pool.connection_tagged("da_dispatcher").await?;
+        let batches = conn
+            .data_availability_dal()
+            .get_ready_for_da_dispatch_l1_batches(self.config.max_rows_to_dispatch() as usize)
+            .await?;
+        drop(conn);
+
+        for batch in batches {
+            let dispatch_latency = METRICS.blob_dispatch_latency.start();
+            let dispatch_response = retry(self.config.max_retries(), batch.l1_batch_number, || {
+                self.client
+                    .dispatch_blob(batch.l1_batch_number.0, batch.pubdata.clone())
```

Comment:
Just to clarify: Do I understand correctly that `dispatch_blob()` is supposed to be idempotent (i.e., if called multiple times with the same params, it should work fine? Could be worth clarifying in the `DataAvailabilityClient` docs.

Reply:
correct, added to the README